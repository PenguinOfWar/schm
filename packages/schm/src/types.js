// @flow

export type Parser = (
  value: any,
  option: any,
  options?: Object,
  values?: Object,
  params?: Object,
) => any

export type ValidatorResponse = {
  valid: any,
  message?: string,
  isSchema?: boolean,
}

export type Validator = (
  value: any,
  option: any,
  options?: Object,
  values?: Object,
  params?: Object,
) => ValidatorResponse

/** */
export type ValidationError = {
  param: string,
  value?: any,
  validator: string,
  message?: string,
}

/** */
export type Schema = {
  params: Object,
  parsers: { [string]: Parser },
  validators: { [string]: Validator },
  parse: (values: Object) => Object,
  validate: (values: Object) => Promise<ValidationError[]>,
  merge: (...schemas: Schema[]) => Schema,
}

/** */
export type SchemaGroup = (previous: Schema) => Schema
