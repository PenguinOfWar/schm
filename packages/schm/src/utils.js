// @flow
import type { Schema } from './types'

type TransformValueFunction = (
  value: any,
  options: Object,
  paramName: string,
  paramPath: string,
) => any

export const toArray = (value: any): any[] => (
  [].concat(typeof value === 'undefined' ? [] : value)
)

export const isSchema = (schema: ?Schema): boolean => !!(
  schema &&
  schema.params &&
  schema.parsers &&
  schema.validators &&
  schema.parse &&
  schema.validate &&
  schema.merge
)

export const mapValuesToSchema = (
  values: Object = {},
  schema: Schema,
  transformValue: TransformValueFunction,
  paramNames: string[] = [],
): Object => (
  Object.keys(schema.params).reduce((finalParams, paramName) => {
    const options = schema.params[paramName]
    const value = values[paramName]
    let finalValue

    if (isSchema(options.type)) {
      finalValue = value ? mapValuesToSchema(
        value,
        options.type,
        transformValue,
        [...paramNames, paramName],
      ) : undefined
    } else if (Array.isArray(options.type)) {
      const arrayValue = toArray(value)

      if (isSchema(options.type[0].type)) {
        finalValue = arrayValue.map((val, i) => mapValuesToSchema(
          val,
          options.type[0].type,
          transformValue,
          [...paramNames, paramName, `${i}`]
        ))

      // RECURSIVE
      // RECURSIVE
      // RECURSIVE
      // RECURSIVE
      // RECURSIVE
      // RECURSIVE
      } else if (Array.isArray(options.type[0].type)) {
        const arrayValue = toArray((value || [])[0])

        if (isSchema(options.type[0].type[0])) {
          finalValue = arrayValue.map((val, i) => mapValuesToSchema(
            val,
            options.type[0].type[0].type,
            transformValue,
            [...paramNames, paramName, `${i}`]
          ))
        } else {
          finalValue = arrayValue.map((val, i) => transformValue(
            val,
            options.type[0].type[0],
            paramName,
            [...paramNames, paramName, i].join('.')
          ))
        }
      } else {
        finalValue = arrayValue.map((val, i) => transformValue(
          val,
          options.type[0],
          paramName,
          [...paramNames, paramName, i].join('.')
        ))
      }
    } else {
      finalValue = transformValue(
        value,
        options,
        paramName,
        [...paramNames, paramName].join('.')
      )
    }

    return {
      ...finalParams,
      [paramName]: finalValue,
    }
  }, {})
)
