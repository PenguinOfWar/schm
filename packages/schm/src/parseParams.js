import { isSchema } from './utils'
import { defaultSchema } from './schema'

const isFunction = options => typeof options === 'function'
const isObject = options => typeof options === 'object'
const hasType = options => isObject(options) && options.type

const literalType = (options) => {
  if (isFunction(options) || isSchema(options)) {
    return { type: options }
  }
  return options
}

const defaultValue = (options) => {
  if (!isObject(options) && !isFunction(options)) {
    return { type: options.constructor, default: options }
  }
  return options
}

const nestedObject = (options) => {
  if (!isSchema(options) && isObject(options) && !hasType(options)) {
    return { type: defaultSchema(options) }
  }
  return options
}

const parseOptions = (options) => {
  if (Array.isArray(options)) {
    return { type: [parseOptions(options[0])] }
  }
  if (hasType(options) && Array.isArray(options.type)) {
    return { type: [parseOptions(options.type[0])] }
  }

  const compose = (...fns) => fns.reduce((f, g) => (...args) => f(g(...args)))
  const flow = compose(literalType, defaultValue, nestedObject)
  return flow(options)
}

const parseParams = params => (
  Object.keys(params).reduce((finalParams, paramName) => ({
    ...finalParams,
    [paramName]: parseOptions(params[paramName]),
  }), {})
)

export default parseParams
