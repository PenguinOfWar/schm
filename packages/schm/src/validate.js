// @flow
import type { Schema, ValidationError } from './types'
import mapValues from './mapValues'
import { parseValidatorOption } from './utils'

const isPrimitive = arg =>
  arg == null || (typeof arg !== 'object' && typeof arg !== 'function')

const replaceMessage = (
  message: string,
  paramName: string,
  value: any,
  validatorName: string
): string => (
  message
    .replace(/\{(PARAM|PATH)\}/g, paramName)
    .replace(/\{VALUE\}/g, value)
    .replace(/\{(VALIDATOR|TYPE)\}/g, validatorName)
)

const createErrorObject = (
  param: string,
  value: any,
  validator: string,
  optionValue: any,
  message?: string
): ValidationError => ({
  param,
  ...isPrimitive(value) ? { value } : {},
  validator,
  ...isPrimitive(optionValue) ? { [validator]: optionValue } : {},
  ...message ? { message: replaceMessage(message, param, value, validator) } : {},
})

/**
 * Validates a schema based on given values.
 * @example
 * const userSchema = schema({
 *   name: {
 *     type: String,
 *     required: true,
 *   },
 *   age: {
 *     type: Number,
 *     min: [18, 'Too young'],
 *   }
 * })
 *
 * validate({ name: 'John', age: 17 }, userSchema)
 *   .then((parsedValues) => {
 *     console.log('Yaay!', parsedValues)
 *   })
 *   .catch((errors) => {
 *     console.log('Oops!', errors)
 *   })
 *
 * \/*
 * Output:
 * Oops! [{
 *   param: 'age',
 *   value: 17,
 *   validator: 'min',
 *   min: 18,
 *   message: 'Too young',
 * }]
 * *\/
 */
const validate = (values?: Object = {}, schema: Schema): Promise<ValidationError[]> => {
  const parsed = schema.parse(values)
  const promises = []
  const errors = []

  mapValues(parsed, schema.params, (value, options, paramName, paramPath) => {
    let error

    Object.keys(options).forEach((optionName) => {
      if (error) return

      const option = parseValidatorOption(options[optionName], optionName === 'enum')
      const validator = schema.validators[optionName]

      if (typeof validator === 'function') {
        const result = validator(value, option, options, parsed, schema.params)
        const { valid, message, isSchema } = result

        if (!valid) {
          error = createErrorObject(
            paramPath,
            value,
            optionName,
            option.optionValue,
            message,
          )
          errors.push(error)
        } else if (typeof valid.catch === 'function') {
          const promise = valid.catch((schemaErrors) => {
            if (isSchema) {
              return errors.push(...schemaErrors)
            }
            const errorObject = createErrorObject(
              paramPath,
              valid,
              optionName,
              option.optionValue,
              message,
            )
            return Promise.reject(errorObject)
          })
          promises.push(promise)
        }
      } else if (validator) {
        throw new Error(`[schm] ${paramName} validator must be a function`)
      }
    })
  })

  return Promise.all(promises).then(() => {
    if (errors.length) {
      return Promise.reject(errors)
    }
    return parsed
  }, (e) => {
    const allErrors = [].concat(errors, e || [])
    return Promise.reject(allErrors)
  })
}

export default validate
