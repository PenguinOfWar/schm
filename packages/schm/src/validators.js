// @flow
import omit from 'lodash/omit'
import { isSchema } from './utils'
import type { Validator } from './types'

type ParsedOption = {
  optionValue: any,
  message?: string,
}

const isArrayAndHasMessage = (option: any): boolean => (
  Array.isArray(option) &&
  option.length === 2 &&
  typeof option[1] === 'string'
)

const isValidatorObject = (option: any): boolean => (
  option && (option.message || option.msg)
)

const parseOption = (option: any, ignoreArray?: boolean): ParsedOption => {
  if (!ignoreArray && isArrayAndHasMessage(option)) {
    return {
      optionValue: option[0],
      message: option[1],
    }
  } else if (isValidatorObject(option)) {
    const optionWithoutMessage = omit(option, 'message', 'msg')
    const optionValueKey = Object.keys(optionWithoutMessage)[0]
    return {
      optionValue: optionWithoutMessage[optionValueKey],
      message: option.message || option.msg,
    }
  }
  return { optionValue: option }
}

export const validate: Validator = (value, option, values, options, params) => {
  const { optionValue, message } = parseOption(option)
  if (Array.isArray(optionValue)) {
    return optionValue.reduce((response, currentOption) => (
      !response.valid ? response : validate(value, currentOption)
    ), { valid: true })
  }
  if (typeof optionValue !== 'function') {
    throw new Error('[schm] validate must be a function')
  }
  return {
    valid: optionValue(value, values, options, params),
    message,
  }
}

export const type: Validator = (value, option) => {
  if (isSchema(option)) {
    return { valid: option.validate(value), isSchema: true }
  }
  return { valid: true }
}

export const required: Validator = (value, option) => {
  const { optionValue, message } = parseOption(option)
  const valid = optionValue
    ? value != null && value !== '' && !Number.isNaN(value)
    : true
  return {
    valid,
    message: message || '{PARAM} is required',
  }
}

export const match: Validator = (value, option) => {
  const { optionValue, message } = parseOption(option)
  if (!(optionValue instanceof RegExp)) {
    throw new Error('[schm] match must be a regex')
  }
  return {
    valid: !value || optionValue.test(value),
    message: message || '{PARAM} does not match',
  }
}

export const enumValidator: Validator = (value, option) => {
  const { optionValue, message } = parseOption(option, true)
  if (!Array.isArray(optionValue)) {
    throw new Error('[schm] enum must be an array')
  }
  return {
    valid: optionValue.indexOf(value) >= 0,
    message: message || `{PARAM} must be one of the following: ${optionValue.join(', ')}`,
  }
}

export const max: Validator = (value, option) => {
  const { optionValue, message } = parseOption(option)
  return {
    valid: typeof value === 'undefined' || value <= optionValue,
    message: message || `{PARAM} must be lower than or equal ${optionValue}`,
  }
}

export const min: Validator = (value, option) => {
  const { optionValue, message } = parseOption(option)
  return {
    valid: typeof value === 'undefined' || value >= optionValue,
    message: message || `{PARAM} must be greater than or equal ${optionValue}`,
  }
}

export const maxlength: Validator = (value, option) => {
  const { optionValue, message } = parseOption(option)
  return {
    valid: typeof value === 'undefined' || value.length <= optionValue,
    message: message || `{PARAM} length must be lower than or equal ${optionValue}`,
  }
}

export const minlength: Validator = (value, option) => {
  const { optionValue, message } = parseOption(option)
  return {
    valid: typeof value === 'undefined' || value.length >= optionValue,
    message: message || `{PARAM} length must be greater than or equal ${optionValue}`,
  }
}

export default {
  type,
  required,
  match,
  enum: enumValidator,
  max,
  min,
  maxlength,
  minlength,
  validate,
}
