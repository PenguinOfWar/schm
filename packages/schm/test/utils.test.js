import schema from '../src'
import { isSchema, mapValuesToSchema } from '../src/utils'

test('isSchema', () => {
  expect(isSchema({})).toBe(false)
  expect(isSchema(schema())).toBe(true)
})

test.only('mapValuesToSchema', () => {
  const schm = schema({
    foo: String,
    bar: {
      baz: [{
        qux: [{
          type: String,
        }],
      }],
    },
  })
  const values = {
    foo: 'foo',
    bar: {
      baz: [
        { qux: ['1', '2'] },
        { qux: '3' },
      ],
    },
  }
  const fn = jest.fn((value, options, paramName, paramPath) => ({
    paramName,
    options,
    value,
    paramPath,
  }))
  expect(mapValuesToSchema(values, schm, fn)).toMatchSnapshot()
})
