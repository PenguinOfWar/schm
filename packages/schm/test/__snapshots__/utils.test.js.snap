// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapValuesToSchema 1`] = `
Object {
  "bar": Object {
    "baz": Array [
      Object {
        "qux": Array [
          Object {
            "options": Object {
              "type": [Function],
            },
            "paramName": "qux",
            "paramPath": "bar.baz.0.qux.0",
            "value": "1",
          },
          Object {
            "options": Object {
              "type": [Function],
            },
            "paramName": "qux",
            "paramPath": "bar.baz.0.qux.1",
            "value": "2",
          },
        ],
      },
      Object {
        "qux": Array [
          Object {
            "options": Object {
              "type": [Function],
            },
            "paramName": "qux",
            "paramPath": "bar.baz.1.qux.0",
            "value": "3",
          },
        ],
      },
    ],
  },
  "foo": Object {
    "options": Object {
      "type": [Function],
    },
    "paramName": "foo",
    "paramPath": "foo",
    "value": "foo",
  },
}
`;
